### ISSUE TITLE
BadImageFormatException on codegeneration in 2.0.0-beta1, targetting netstandard2.0 win7-x64

### ISSUE THREAD

**Issue by rikbosch** [dotnet/orleans#3619]
Hi,

When trying to build a grain implementations project, targetting netstandard2.0, runtime win7-x64 I get the following exception:

``` 
5>Orleans-CodeGen - Generating file D:\Projects\AO\src\api\Ao.Core.Grains\obj\Debug\netstandard2.0\win7-x64\Ao.Core.Grains.orleans.g.cs
5>warn: Orleans.CodeGenerator.SerializerGenerationManager[103807]
5>      System type ClaimsPrincipal may require a custom serializer for optimal performance. If you use arguments of this type a lot, consider submitting a pull request to https://github.com/dotnet/orleans/ to add a custom serializer for it.
5>-- Code Generation FAILED --
5>
5>Exc level 0: System.BadImageFormatException: Could not load file or assembly 'System.Data.SqlClient, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context. (Exception from HRESULT: 0x80131058)
5>   at System.Signature.GetSignature(Void* pCorSig, Int32 cCorSig, RuntimeFieldHandleInternal fieldHandle, IRuntimeMethodInfo methodHandle, RuntimeType declaringType)
5>   at System.Reflection.RuntimeConstructorInfo.get_Signature()
5>   at System.Reflection.RuntimeConstructorInfo.GetParametersNoCopy()
5>   at System.RuntimeType.FilterApplyMethodBase(MethodBase methodBase, BindingFlags methodFlags, BindingFlags bindingFlags, CallingConventions callConv, Type[] argumentTypes)
5>   at System.RuntimeType.GetConstructorCandidates(String name, BindingFlags bindingAttr, CallingConventions callConv, Type[] types, Boolean allowPrefixLookup)
5>   at System.RuntimeType.GetConstructorImpl(BindingFlags bindingAttr, Binder binder, CallingConventions callConvention, Type[] types, ParameterModifier[] modifiers)
5>   at Orleans.CodeGenerator.SerializerGenerator.GetObjectCreationExpressionSyntax(Type type)
5>   at Orleans.CodeGenerator.SerializerGenerator.GenerateDeepCopierMethod(Type type, List`1 fields)
5>   at Orleans.CodeGenerator.SerializerGenerator.GenerateClass(String className, Type type, Action`1 onEncounteredType)
5>   at Orleans.CodeGenerator.RoslynCodeGenerator.GenerateForAssemblies(Assembly targetAssembly, Boolean runtime)
5>   at Orleans.CodeGenerator.RoslynCodeGenerator.GenerateSourceForAssembly(Assembly input)
5>   at Orleans.CodeGeneration.CodeGenerator.GenerateSourceForAssembly(Assembly grainAssembly)
5>   at Orleans.CodeGeneration.CodeGenerator.GenerateCodeInternal(CodeGenOptions options)
5>   at Orleans.CodeGeneration.CodeGenerator.GenerateCode(CodeGenOptions options)
5>   at Orleans.CodeGeneration.Program.Main(String[] args)
5>Exc level 1: System.BadImageFormatException: Cannot load a reference assembly for execution.
5>C:\Users\r.bosch\.nuget\packages\microsoft.orleans.orleanscodegenerator.build\2.0.0-beta1\build\Microsoft.Orleans.OrleansCodeGenerator.Build.targets(69,5): error MSB3073: The command ""dotnet" "C:\Users\r.bosch\.nuget\packages\microsoft.orleans.orleanscodegenerator.build\2.0.0-beta1\build\..\tasks\netcoreapp2.0\Orleans.CodeGeneration.Build.dll" "@obj\Debug\netstandard2.0\win7-x64\codegen\Ao.Core.Grains.orleans.g.args.txt"" exited with code 3.
5>Done building project "Ao.Core.Grains.csproj" -- FAILED.
========== Build: 4 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
``` 

targetting net462, runtime win7-x64 it succeeds, with warnings:

```
5>Orleans-CodeGen - Generating file D:\Projectens\AO\src\api\Ao.Core.Grains\obj\Debug\net462\win7-x64\Ao.Core.Grains.orleans.g.cs
5>warn: Orleans.CodeGenerator.SerializerGenerationManager[103807]
5>      System type ClaimsPrincipal may require a custom serializer for optimal performance. If you use arguments of this type a lot, consider submitting a pull request to https://github.com/dotnet/orleans/ to add a custom serializer for it.
5>warn: Orleans.CodeGenerator.RoslynCodeGenerator[101723]
5>      Exception loading types from assembly 'Core.Services, Version=0.1.0.0, Culture=neutral, PublicKeyToken=null':
5>      Exc level 0: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
5>         at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
5>         at System.Reflection.RuntimeAssembly.get_DefinedTypes()
5>         at Orleans.Runtime.TypeUtils.GetDefinedTypes(Assembly assembly, Logger logger) in D:\build\agent\_work\18\s\src\Orleans.Core\CodeGeneration\TypeUtils.cs:line 555
5>      Exc level 1: System.BadImageFormatException: Could not load file or assembly 'System.Data.SqlClient, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context. (Exception from HRESULT: 0x80131058)
5>      Exc level 2: System.BadImageFormatException: Cannot load a reference assembly for execution..
5>System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
5>   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
5>   at System.Reflection.RuntimeAssembly.get_DefinedTypes()
5>   at Orleans.Runtime.TypeUtils.GetDefinedTypes(Assembly assembly, Logger logger) in D:\build\agent\_work\18\s\src\Orleans.Core\CodeGeneration\TypeUtils.cs:line 555
5>Orleans-CodeGen - Generated file written D:\Projectens\AO\src\api\Ao.Core.Grains\obj\Debug\net462\win7-x64\Ao.Core.Grains.orleans.g.cs
5>Ao.Core.Grains -> D:\Projectens\AO\src\api\Ao.Core.Grains\bin\Debug\net462\win7-x64\Ao.Core.Grains.dll
========== Build: 5 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by attilah** [comment]
@rikbosch Hi! I tried to reproduce it, and I was not able to have a modified version HelloWorld.NetCore which fails to build. Please provide more details, or a small repro project would be even better.

**Comment by attilah** [comment]
@rikbosch Also this file would maybe help: "@obj\Debug\netstandard2.0\win7-x64\codegen\Ao.Core.Grains.orleans.g.args.txt" please attach it

**Comment by rikbosch** [comment]
@attilah 
here are the contents of the file:

https://gist.github.com/rikbosch/f2d75cff5619552c478e6a78453ab20a

**Comment by rikbosch** [comment]
I've been able to reproduce the issue (on my machine ™️ ) in the HelloWorld solution

I've pushed the code here: https://github.com/rikbosch/orleans/tree/issues/3619_badimagexception_on_codegen

**Comment by attilah** [comment]
@rikbosch Thanks for the repro, I was able to use it and create a PR.

The logs you see were just type loading warnings and not errors, types are getting generated for that assembly. I enhanced the warning logging method to only output when we encounter any other type of exception, but not BadImageFormatException.

Thanks for reporting.

**Comment by attilah** [comment]
@rikbosch however the repro worked for the second stacktrace (even with netstandard 2.0 target), I was not able to repro the first stacktrace, which is completely different, can you still repro that one too?

**Comment by rikbosch** [comment]
@attilah that's actually right. I cannot reproduce the exception using the example project.

I have, however narrowed it down by debugging the code generator itself.

There is a serializer generated for a class which is used in the generic definition of a grain implementation.

this class has 1 constructor, with an SqlConnection parameter.

the line which throws in the exception is: `typeInfo.GetConstructor(Type.EmptyTypes) != null`

https://github.com/dotnet/orleans/blob/master/src/Orleans.CodeGeneration/SerializerGenerator.cs#L423

What I'm trying to figure out is, why a serializer is being generated for this type, as it is not used in any Grain methods or State objects.

**Comment by attilah** [comment]
@rikbosch thanks for the additional info, I was able to repro it this way, I updated my PR with a fix.

I don't know why your state class in ```Grain<T>``` needs a sql connection, generally I think persistence would happen at the grain level and ```<T>``` should be a POCO or alike.

On the other hand in your situation, you can have an ```IDbConnection``` there which is not causing any problem with the currently released 2.0.0-beta1 bits.
